### DISCUSSION TITLE
Parallel IO

### DISCUSSION THREAD

**Discussion by pavlis** [mspass-team/mspass#444] [Design & Development]
In the process of reworking readers and writers in the Database class I have needed to consider the related problem of how we can do reads and writes faster with parallel io systems.   Currently, our best attempt lives in two functions called `read_distributed_data` and `write_distributed_data`.   Looking at them, I have some ideas for how we might rework both functions to allow them do IO faster in parallel.  

An important insight is something I gleaned from cleaning up the data object writer in the `Database` class is to recognize that our atomic data objects need to be treated as four distinctly different types of information:

1. The sample data - stored as a vector or matrix container. 
2. The thing we call `Metadata` that == python dictionary (almost).   
3. `ErrorLogger` data
4. `ProcessingHistory` data

We currently treat save/read just that way partitioning each piece to separate places.  

1. Sample data are stored either as files or in gridfs
2. `Metadata` contents are pushed to the appropriate wf collection.
3. `ErrorLogger` data are pushed to an "elog" collection (currently).  
4. `ProcessingHistory` data is optional, but when defined is pushed to another collection.  I think it is called "history".   

Note there is a special case for saving data: what to do with data marked dead.   We properly normally preserve a record of data killed as documents in the "elog" collection.   I started a related discussion page on this issue yesterday and suggested we should put bodies in a different collection.   I recommended it be called "cemetery".   

With that review of status, let's make clear what it is we need to solve here:
1.   We need to provide cleaner support for parallel IO as we have seen many examples of IO bound workflows.  We need to remember an alternative definition of a supercomputer is a computer that turns a compute bound job to an io bound job.   
2. Database transaction in a parallel constructs are always problematic for performance without heroic effort.  There are always latency issues with the MongoDB server and if a lot of processes yell at the server simultaneously it could get overwhelmed. 
3. Passing around database handles between nodes is intrinsically dangerous.   @wangyinz  did something magical to make it work so far, but I am far from convinced there aren't situations where that model could fail.   The reason, of course, is moving any IO handle around is impossible to dangerous.    A C FILE pointer, for example, could never be passed between nodes since the handle has to be able to see the io buffers hidden by the api.  MongoDB seems to have been designed for parallel environments so maybe that is why ths works, but still seems we should try to avoid serializing things like a MongoDB cursor.  

That said, I have two suggestions:
1.  To get the most bang for the buck with a parallel file system like HDF5 we should probably limit use to the sample data.   That is almost always the biggest volume of data to be saved and would seem to be the most important to optimize. 
2. I think we can improve reads and writes of the other 3 components noted above with one simple addition:  Do not use the `elog` and `history` collections in a parallel setup but keep the same information as subdocuments in a waveform collection.   Writers would save that information and readers could normally chose to discard the elog and history subdocuments. 

The above to suggestions allow this algorithm for parallel save using an input bag/rdd:

1. Save sample data with extensions for storage modes like hdf5.   Writers would return an edited copy of the data received with attributes sufficient to later reconstruct the sample data (we more or less do this now but without a parallel interface).    Note because the input is 
a bag/rdd this would be constructed as a map call inside the function. 
2.  Create a python dict from the Metadata, elog, and (optional) history data.   In this model the elog and history data would be subdocument with appropriate tags while the Metadata is handled exactly as now (note that include the schema constraints).   That operation could be done in a single function that takes a mspass object as input and returns the python dictionary with schema constraints applied and any elog or history data. 
3. The database save could probably be done in parallel with a map_partition with a new Database handle created by partition.   That should be always safe.  

A reader could use a similar model but more or else reversed.  
1.  We start with a parallel reader to create a bag/rdd from a wf collection.   We have the prototype loading such data into a dataframe and then generating a bag from the dataframe. I think we need more work on how to do that most efficiently, but that is a necessary evil, we know, because otherwise one is subject to cursor timeout issues as well as the problem of dask or spark trying to serialize a cursor.   Point is somehow we need to most efficiently form a bag/rdd of wf documents with or without data from normalization collections.   The details of that are yet to be fully worked out. 
2. Use the bag of dictionaries to drive  parallel readers that construct data objects and load the sample data.   That is a nice devision for different parallel io systems as the detailed reads could depend upon storage modes beyond the generic "file".

**Comment by pavlis** [comment]
I did some experimentation with dask's `map_partitions` bag method.   The documentation for that method is really cryptic, but experiments with toy problems helped me establish a few key things the documentation obscured (at least to me):

1.  A function to be used with `map_partitions` needs to expect arg0 to be a list of the data of the type of each bag component.  
2. The list is a length computed from the size of the bag and the number of partitions.  e.g. if the bag size is 10000 and the number of partitions is 10 each call to the function can expect to receive a list of 1000 items to process.   
3. That means, of course, that the basic difference compared to a function used with map is the algorithm has to have a loop over the list instead of handling one thing (what a map expects).   The documentation describes this with the cryptic statement that the input "must be an interable", which is ridiculously confusing and actually misleading.  `map_partition` does the chopping up of the container into pieces and is more or less like calling compute on each chunk.
4.  Quick scans of the documentation show pyspark has a [similar function](https://spark.apache.org/docs/latest/api/python/reference/api/pyspark.RDD.mapPartitions.html)  but I've not played with it for any comparison.   The docs there suggest it behaves pretty much the same way.   

From that experience I produced the following, completely untested, prototype for a revised version of `write_distributed_data`.   It uses reworked methods in `Undertaker` and `Database` I will discuss below but for now here is the draft with the docstrings stripped:
```
def partioned_save_wfdoc(docs,
                         collection="wf_TimeSeries",
                         db=None,
                         dbname=None,
        ):
    if db is None:
        dbclient = DBClient()
        # needs to throw an exception of both db and dbname are none
        db = dbclient.get_database(dbname)
    dbcol = db[collection]
    wfids = dbcol.insert_many(docs).inserted_id()
    return wfids

def write_distributed_data(
    data,
    db,
    mode="promiscuous",
    storage_mode="gridfs",
    format=None,
    file_format=None,
    overwrite=False,
    exclude_keys=None,
    collection=None,
    data_tag=None,
    post_elog=True,
    post_history=False,
    alg_name="write_distributed_data",
    alg_id="0",
):

    #TODO  Needs to handle schema and collection based on type of input data
    # or via args
    stedronsky=Undertaker(db)
    # Note although this is a database method it will not refernce 
    # db at all if writing to files
    # also note storage_mode allows specifying a parallel file system
    # dir and dfile set None is assumed to be bombproof.  Current 
    # implementation tries to extract value form metadata and if it 
    # isn't defined there it will create something - see docstring 
    data = data.map(db._save_sample_data,
                    storage_mode=storage_mode,
                    dir=None,
                    dfile=None,
                    format=format,
                    overwrite=overwrite,
                    )
    data = data.map(stedronsky.mummify,post_elog,post_history)
    data = data.map(md2doc,save_schema,exclude_keys,mode,post_elog, post_history)
    data = data.map_partition(partioned_save_wfdoc,db=db,collection=collection)

    if format == "spark":
        data = data.collect()  # rdd -> list
    else:
        data = data.compute()  # bag -> list
    return data
```
Key points:

- The first function definition is an example of one that would use map_partitions.  It illustrates a key reason for why that approach could be useful.  That is, this implementation would allow the option of having the function construct a new database handle for each partition.   The safest, and likely slowest, algorithm would be to have a new handle be created for each partition.   The fastest would serialize the db handle, but as we know that could be dangerous.   Likely the optimal solution is to exploit sharding and have every node contain a shard that all worker threads running on that node would use the local shard.   I haven't a clue how to implement that, but it is worth stating at this design phase.   For now, I think the algorithm above should work, but I emphasize it is NOT tested and far from complete (e.g. lacks critical arg checking)
- `db._save_sample_data` is a new method of `Database` that standardizes saving the sample data.   A db handle is only needed if storage_mode is set to "gridfs".   That means it should work for files even if the database handle doesn't serialize.  More importantly, by regularizing the function for saving the sample data we should be able to more easily add support for a parallel file system.  The idea is that we will only need to add a private method that would write data for a different "storage_mode".  e.g. we could produce an hdf5 writer and allow support for `storage_mode="hdf5"`.    I may be way off base on that so correct me.  I am not well versed in the hdf5 or other parallel io systems.
- This prototype handles dead data with the revised `Undertaker` class discussed in a related page.   It calls the `mummify` method with the `post_elog` and `post_history` arguments.  These are booleans that when set True cause the contents of ErrorLogger and ProcessingHistory to be posted to Metadata as subdocuments with keys "error_log" and "processing_history" respectively.  The only real difference is that mummify takes a mspass data object in and emits a mspass data object of the same type.  Dead data are "muffified" which means only that we call the `set_npts` method and set the sample array size to 0 for all dead data.  The mummify method  takes a mspass data object in and emits a mspass data object of the same type. 
- The call to `md2doc` is another new function in the revised db.py module.  It standardizes the complex code that imposes schema constraints and converts the contents of a Metadata container to a python dictionary that can be pushed to MongoDB.  Hence, the output of the ms2doc is a bag of python dictionaries (aka documents) that can be pushed directly to the database.   Note this function has to handle elog and history like mummify but it ignores dead data and just converts the Metadata container for such data.  (TODO NOTE:  Draft code in a branch not yet pushed does not handle this issue right.   Need a standard way to handle how elog and history are saved for live and/or dead data.  Functions need to be usable by both this parallel interface and the Database class and there are implementation details to be worked out. )
- The special function at the top of above used in the `map_partitions` call is pretty simple.  Because `md2doc` returns a bag of python dictionaries that `map_partitions` blocks into an array of lists passed to the function, we can pass that list directly to MongoDB's `insert_many` method.  That line, is completely untested but it illustrates the concept.  I close by pointing out that using insert_many like this should significantly reduce database latency time.  You reduce the number of transactions from the total number of wf documents to the number of partitions.   That may make the high cost of constructing a new handle for each partition justified for the stability it might buy.   That is a big unknown, however, that we need to study.  I'm not sure how to do that.

**Comment by pavlis** [comment]
I have written a complete, new implementation of `read_distributed_data` cleaning up some of the redundant junk and meshing it with the revisions to `Database.read_data`.  I followed the prototype written a few months ago by having the function first convert input defined by a Database handle to a Pandas DataFrame.   I know the reason for that original choice was to allow direct input by any implementation of DataFrame (e.g. dask's variant).  We definitely want that functionality, but my first draft led to this section for handling atomic data (Note the prototype didn't even support ensembles which have to be handled completely differently.   I'll post that in a second box next.): 
```
        # convert dask dataframe to pandas dataframe to mesh with dask and spark api
        if isinstance(data, daskDF):
            data = data.compute()

        # convert spark dataframe to spark dataframe
        elif isinstance(data, sparkDF):
            data = data.toPandas()

        # now the type of data is a pandas dataframe
        if format == "spark":
            plist = spark_context.parallelize(
                data.to_dict("records"), numSlices=npartitions
            )
            plist = plist.map(lambda doc : db.read_data(
                                    doc,
                                    collection=collection,
                                    normalize=normalize,
                                    load_history=load_history,
                                    exclude_keys=exclude_keys,
                                    data_tag=data_tag,
                                    aws_access_key_id=aws_access_key_id,
                                    aws_secret_access_key=aws_secret_access_key,
                                )
                            )
        else:
            plist = daskbag.from_sequence(data.to_dict("records"), npartitions=npartitions)
            plist = plist.map(
                db.read_data,
                collection=collection,
                normalize=normalize,
                load_history=load_history,
                exclude_keys=exclude_keys,
                data_tag=data_tag,
                aws_access_key_id=aws_access_key_id,
                aws_secret_access_key=aws_secret_access_key,
                )
        return plist
```
What I realized is dumb about this algorithm is that the internal version of the DataFrame to a pandas 
DataFrame is not scalable.   Although not yet fully tested, the last section running `read_data` in 
parallel for dask and spark is what we want and is scalable.  I need help from collaborators who were 
involved in the prototype to explain to me why the pandas conversion above below the 
comment "convert dask dataframe to pandas dataframe to mesh with dask and spark api" is necessary?   

Perhaps this mismatch partly came about by the change I made to the api for `Database.read_data`.   It now 
expects mainly to receive a python dict as input while the previous version wanted an ObjectID.  As noted elsewhere that caused an unnecessary inefficiency.   It seems to me the natural input for a parallel reader is not a 
DataFrame but a generalized list (i.e. bag or RDD) of python dict containers loaded from MongoDB.   
We can still allow DataFrame input but recognize that is mainly a format convenience to allow loading of data 
through some other mechanism like a relational database.   DataFrame input could be handled similar to  
above where what is doing anyway is converting the dataframe tuples to dict containers on the fly.   

I'm kind of stuck on this implementation until we can clear this detail.   I'm not exactly sure how to load a bag/RDD directly from MongoDB without generating a similar scalability issue.  e.g. with dask you can't just use the "from_sequence" method of bag as the sequence has to be the entire list of Metadata.  We need a scalable way to do that.  Any ideas?

**Reply by wangyinz** [reply]
For the conversion to Pandas dataframe, I think the reason of having that is that we wanted to support input of any dataframe implementations in this function. I don't think scalability is something we considered here because the input should mostly (or always) be a list of metadata. We recognized that the list of metadata may be in the form of dataframes, so we take that as the input, eventually convert it into a list, and then do the distributed read. 

I guess the change in `read_data` is indeed making it incompatible with this function now. Basically, the previous implementation reads a single data object at a time. Then, we can call that with a map method in this function so that individual task is a separate read (including db operations and reading wf files), and it happens on the worker, and we will be getting the result distributed across the workers avoiding the scaling issue.

**Reply by pavlis** [reply]
I'll see if I can figure out a way to reduce the memory footprint to produce a bag of docs.  I know of one, but I think it might be much slower.   That is, we could produce a list of ObjectIds, which would always be about as small as possible, and the use a query-driven map function.   That is more or less the way the old version worked.   If I can't think of anything else I could easily implement the option to allow the input to be a list of ObjectIds and have the documentation suggest using that version if the database input causes a memory fault.   

Not sure what you are suggesting in paragraph 2.  This version still reads one atomic datum at a time but is driven by a document.  The read is in a map call but uses the new version of read_data.   Because this version uses a dataframe as an intermediary both the dask and spark versions have to convert the dataframe tuples to dict containers that are passed to the revised reader.   That intermediate step makes scalability worse as effectively two formats have to coexist.   Writing this, however, gave me a dahh moment.  A dask or pyspark dataframe can be used directly as above but should just not be converted to a pandas DataFrame.   I just check the documentation and both pyspark and dask implement the "to_dict" method of DataFrame.   Hence, we should just advertise the dask or pyspark DataFrame input as the choice for enormous datasets.  I'm going to go that way for now anyway.  Let me know if you think otherwise.

**Reply by wangyinz** [reply]
Oh, I was thinking that the new implementation reads a list now. Probably I got confused with your other implementation. Anyway, I don't see a to_dict method for the dask and pyspark variant of DataFrame in either of the documentation. It actually makes no sense to convert that into a dict directly because the DataFrame in those frameworks are distributed objects. I guess if we do want to avoid scaling issue, we can convert them directly into Bag and RDD with the [to_bag](https://docs.dask.org/en/latest/generated/dask.dataframe.DataFrame.to_bag.html#dask.dataframe.DataFrame.to_bag) method and the [rdd](https://spark.apache.org/docs/3.1.1/api/python/reference/api/pyspark.sql.DataFrame.rdd.html#pyspark.sql.DataFrame.rdd) property.

**Comment by pavlis** [comment]
Part 2:  Ensembles

The prototype `read_distributed_data` had not direct support for ensembles.  I devised a reader that I think should preform well, although that is only currently theoretical as we have no tests for this case yet.   

What I did was allow input to be a list of queries == python dict containers.  e.g. some version of the 
following might be used to build a parallel container of ensembles for common source gathers:
```
 [{"source_id" : ObjectId('64b917ce9aa746564e8ecbfd')}, 
            {"source_id" : ObjectId('64b917d69aa746564e8ecbfe')}, 
            ... ]
```
where as usual "..." means there are many of these.   That allows parallel reads with this function:
```
def read_ensemble_parallel(query,
                           db,
                           collection="wf_TimeSeries",
                           mode="promiscuous",
                           normalize=None,
                           load_history=False,
                           exclude_keys=None,
                           data_tag=None,
                           aws_access_key_id=None,
                           aws_secret_access_key=None,
                           ):
    """
    Special function used in read_distributed_data to handle ensembles. 
    
    Ensembles need to be read via a cursor which is not serializable.  
    Here we query a Database class, which is serializable, and 
    call it's read_data method to construct an ensemble that it returns. 
    Defined as a function instead of using a lambda due largely to the 
    complexity of the argument list passed to  read_data.
    """
    cursor = db[collection].find(query)
    ensemble = db.read_data(cursor,
                            collection=collection,
                            mode=mode,
                            normalize=normalize,
                            load_history=load_history,
                            exclude_keys=exclude_keys,
                            data_tag=data_tag,
                            aws_access_key_id=aws_access_key_id,
                            aws_secret_access_key=aws_secret_access_key,
                            )
    kill_me = True
    for d in ensemble.member:
        if d.live:
            kill_me = False
            break
    if kill_me:
        ensemble.kill()
    return ensemble
```
Driven by the following logic in `read_distributed_data` for ensemble input defined by a list of python dicts:
```
        if format == "spark":
            plist = spark_context.parallelize(data, numSlices=npartitions)
            plist = plist.map(lambda q : read_ensemble_parallel(
                                        q,
                                        db,
                                        collection=collection,
                                        normalize=normalize,
                                        load_history=load_history,
                                        exclude_keys=exclude_keys,
                                        data_tag=data_tag,
                                        aws_access_key_id=aws_access_key_id,
                                        aws_secret_access_key=aws_secret_access_key,
                                        )
                                )
        else:
            plist = daskbag.from_sequence(data, npartitions=npartitions)
            plist = daskbag.map(read_ensemble_parallel,
                    db,
                    collection=collection,
                    normalize=normalize,
                    load_history=load_history,
                    exclude_keys=exclude_keys,
                    data_tag=data_tag,
                    aws_access_key_id=aws_access_key_id,
                    aws_secret_access_key=aws_secret_access_key,
                    )
```
Unlike the atomic data problem discussed above the above is completely scalable.  The only item ever created 
as a pure in-memory container is the list of queries.  For any conceivable example I can think of that list should 
be small and never present a scalability issue.  The partitioning/slicing is defined by the conversion to the bag/RDD at the top of the dask/spark blocks.   The task dask/spark are required to execute in this algorithm is reading one ensemble per call.  That means dask/spark should be able to parallelize this read as that atomic "task".   Performance will depend on speed of the reader. 

Comments on this design?

**Reply by wangyinz** [reply]
This design sounds all good to me. I think the tricky part will be on how to build the initial queries. The example above is straightforward, but I can see that some complex queries can also be supported by this design. Maybe we will think of a way to help building the queries, and that will be in a different function.

**Comment by pavlis** [comment]
Yes, I can imagine building more complex queries could get tricky.   On the other hand, the main ensembles I can see people wanting are easily characterized by Ids.   If the existing ones don't work, one could always create a special id tag and update the database with some nonstandard cross-reference/grouping key.

**Comment by pavlis** [comment]
Based on the above, I created the prototype for discussion below.   I lacks complete error handling AND there is a section I haven't quite figure out how to do yet.  I haven't played with pyspark at all so I'm unsure of the exact incantation need in a section that reads a json file to load an RDD.   

Let me explain the key thing change here:
1.  Now all dataframe input is handled directly.   There is no pandas conversion that is not scalable. 
2. For direct database input there are two options.  The default, which is signaled by have "scratchfile=None", uses an in-memory version which loads a  (potentially giant) list of docs (python dict).  When scratchfile is defined the documents are written to a scratchfile and loaded through parallel readers for json files that are a available in both pyspark and dask.   That method allows the documents to be loaded into a bag/RDD.
3. The read section is now common to all input options and is driven by a bag of documents created through all inputs.   
Here is is without the docstring:
```
def read_distributed_data(
    data,
    db=None,
    query=None,
    scratchfile=None,
    collection="wf_TimeSeries",
    mode="promiscuous",
    normalize=None,
    load_history=False,
    exclude_keys=None,
    format="dask",
    npartitions=None,
    spark_context=None,
    data_tag=None,
    aws_access_key_id=None,
    aws_secret_access_key=None,
):
# This is a base error message that is an initialization for 
    # any throw error.  We first two type checking of arg0
    message = "read_distributed_data:  "
    if isinstance(data,list):
        ensemble_mode=True
        i = 0
        for x in list:
            if not isinstance(x,dict):
                message += "arg0 is a list, but component {} has illegal type={}\n".format(i,str(type(x)))
                message += "list must contain only python dict defining queries that define each ensemble to be loaded"
                raise TypeError(message)
    elif isinstance(data,Database):
        ensemble_mode = False
        dataframe_input = False
        db = data
    elif isinstance(data,(pd.Dataframe, sparkDF, daskDF)):
        ensemble_mode = False
        dataframe_input = True
        if isinstance(db,Database):
            db = db 
        else:
            if db:
                message += "Illegal type={} for db argument - required with dataframe input".format(str(type(db)))
                raise TypeError(message)
            else:
                message += "Usage error. An instance of Database class is required for db argument when input is a dataframe"
                raise TypeError(message)
    else:
        message += "Illegal type={} for arg0\n".format(str(type(data)))
        message += "Must be a Dataframe (pandas, spark, or dask), Database, or a list of query dictionaries"
        raise TypeError(message)
        
    # This has a fundamentally different algorithm for handling 
    # ensembles than atomic data.  Ensembles are driven by a list of 
    # queries while atomic data are driven by a dataframe.
    # the dataframe is necessary in this context because MongoDB 
    # cursors can not be serialized while Database can.
    if ensemble_mode:
        if format == "spark":
            plist = spark_context.parallelize(data, numSlices=npartitions)
            plist = plist.map(lambda q : read_ensemble_parallel(
                                        q,
                                        db,
                                        collection=collection,
                                        normalize=normalize,
                                        load_history=load_history,
                                        exclude_keys=exclude_keys,
                                        data_tag=data_tag,
                                        aws_access_key_id=aws_access_key_id,
                                        aws_secret_access_key=aws_secret_access_key,
                                        )
                                )
        else:
            plist = daskbag.from_sequence(data, npartitions=npartitions)
            plist = daskbag.map(read_ensemble_parallel,
                    db,
                    collection=collection,
                    normalize=normalize,
                    load_history=load_history,
                    exclude_keys=exclude_keys,
                    data_tag=data_tag,
                    aws_access_key_id=aws_access_key_id,
                    aws_secret_access_key=aws_secret_access_key,
                    )
        
    else:
        # Logic here gets a bit complex to handle the multiple inputs 
        # possible for atomic data.  That is, multiple dataframe 
        # implementations and a direct database reading mechanism. 
        # The database instance is assumed to converted to a bag of docs
        # above.  We use a set of internal variables to control the 
        # input block used. 
        # TODO:  clean up symbol - plist is an awful name
        if dataframe_input:
            if format == "spark":
                plist = spark_context.parallelize(
                    data.to_dict("records"), numSlices=npartitions
                )
            else:
                plist = data.to_bag()
        else:
            # logic above should guarantee data is a Database 
            # object that can be queried and used to generate the bag/rdd
            # needed to read data in parallel immediately after this block
            if query:
                cursor=data[collection].find(query)
            else:
                cursor=data[collection].find({})
            if scratchfile:
                # here we write the documents all to a scratch file in 
                # json and immediately read them back to create a bag or RDD
                with open(scratchfile, "w") as outfile:
                    for doc in cursor:
                        json.dump(doc,outfile)
                # Intentionally omit error handler here.  Assume 
                # system will throw an error if file open files or write files
                # that will be sufficient for user to understand the problem.
                if format == "spark":
                    # the only way I could find to load json data in pyspark 
                    # is to use an intermediate dataframe.   This should 
                    # still parallelize, but will probably be slower
                    # if there is a more direct solution should be done here.
                    plist = pyspark.read.json(scratchfile)
                    # this is wrong above also don't see how to do partitions
                    plist = plist.map(to_dict,"records")
                else:
                    plist=daskbag.read_text(scratchfile).map(json.loads)
                # TODO delete scratchfile
            else:
                doclist=[]
                for doc in cursor:
                    doclist.append(doc)
                plist = daskbag.from_sequence(doclist)
                del doclist

        # Earlier logic make list a bag/rdd of docs - above converts dataframe to same     
        if format == "spark":
            plist = plist.map(lambda doc : db.read_data(
                                    doc,
                                    collection=collection,
                                    normalize=normalize,
                                    load_history=load_history,
                                    exclude_keys=exclude_keys,
                                    data_tag=data_tag,
                                    aws_access_key_id=aws_access_key_id,
                                    aws_secret_access_key=aws_secret_access_key,
                                )
                            )
        else:
            plist = plist.map(
                db.read_data,
                collection=collection,
                normalize=normalize,
                load_history=load_history,
                exclude_keys=exclude_keys,
                data_tag=data_tag,
                aws_access_key_id=aws_access_key_id,
                aws_secret_access_key=aws_secret_access_key,
                )
    
        return plist
```
